## Final Project: Foundation of Computer Science

2025-26 Project
You have to work on the Trending YouTube dataset.

Notes
It is mandatory to use GitHub for developing the project.
The project must be a jupyter notebook.
There is no restriction on the libraries that can be used, nor on the Python version.
All questions on the project must be asked in the Discussion forum on the course website.
At most 3 students can be in each group. You must create the groups by yourself. You can use the Discussion forum to create the groups.
You do not have to send me the project before the discussion.
You do not have to prepare any slides for the discussion.
You can use AI tools, but you have to describe in the notebook how you have used such tools and you have to show that you have fully understood everything that you have in your project.

--------------------------------

In [1]:
import pandas as pd
import re
import json
import zstandard as zstd
import io
import numpy as np

In [2]:
countries = ["CA", "DE", "FR", "GB", "IN", "JP", "KR", "MX", "RU", "US"]
dfs = {}
for element in countries:
    path = fr"C:\Users\chiar\OneDrive\Desktop\foundation of computer science\trendingYT\{element}videos.csv.zst"
    print("Reading:", path)

    with open(path, "rb") as f: # "rb" opens the file in "read binary" mode, it is necessary because the files ar compressed
        dctx = zstd.ZstdDecompressor()
        decompressed = dctx.decompress(f.read()) # decompression using zstd library

        text = decompressed.decode("utf-8", errors="replace") # decoding with utf-8 that is suitable for all lenguages and special characters

        dfs[element] = pd.read_csv(io.StringIO(text)) # dictionary with all datasets

    dfs[element]["country"] = element # adding the "country" column

Reading: C:\Users\chiar\OneDrive\Desktop\foundation of computer science\trendingYT\CAvideos.csv.zst
Reading: C:\Users\chiar\OneDrive\Desktop\foundation of computer science\trendingYT\DEvideos.csv.zst
Reading: C:\Users\chiar\OneDrive\Desktop\foundation of computer science\trendingYT\FRvideos.csv.zst
Reading: C:\Users\chiar\OneDrive\Desktop\foundation of computer science\trendingYT\GBvideos.csv.zst
Reading: C:\Users\chiar\OneDrive\Desktop\foundation of computer science\trendingYT\INvideos.csv.zst
Reading: C:\Users\chiar\OneDrive\Desktop\foundation of computer science\trendingYT\JPvideos.csv.zst
Reading: C:\Users\chiar\OneDrive\Desktop\foundation of computer science\trendingYT\KRvideos.csv.zst
Reading: C:\Users\chiar\OneDrive\Desktop\foundation of computer science\trendingYT\MXvideos.csv.zst
Reading: C:\Users\chiar\OneDrive\Desktop\foundation of computer science\trendingYT\RUvideos.csv.zst
Reading: C:\Users\chiar\OneDrive\Desktop\foundation of computer science\trendingYT\USvideos.csv.zst


---------------------------

#### 1. Create a single dataframe with the concatenation of all input csv files, adding a column called country

In [4]:
df_all = pd.concat(dfs.values(), ignore_index=True) # concatenation of all datasets from the dictionary
df_all.head()


video_id trending_date  \
0  n1WpP7iowLc      17.14.11   
1  0dBIkQ4Mz1M      17.14.11   
2  5qpjK5DgCt4      17.14.11   
3  d380meD0W0M      17.14.11   
4  2Vv-BfVoq4g      17.14.11   

                                               title channel_title  \
0         Eminem - Walk On Water (Audio) ft. Beyoncé    EminemVEVO   
1                      PLUSH - Bad Unboxing Fan Mail     iDubbbzTV   
2  Racist Superman | Rudy Mancuso, King Bach & Le...  Rudy Mancuso   
3                           I Dare You: GOING BALD!?      nigahiga   
4        Ed Sheeran - Perfect (Official Music Video)    Ed Sheeran   

   category_id              publish_time  \
0           10  2017-11-10T17:00:03.000Z   
1           23  2017-11-13T17:00:00.000Z   
2           23  2017-11-12T19:05:24.000Z   
3           24  2017-11-12T18:01:41.000Z   
4           10  2017-11-09T11:04:14.000Z   

                                                tags     views    likes  \
0  Eminem|"Walk"|"On"|"Water"|"Aftermath/Shady/In...  17158579   787425   
1  plush|"bad unboxing"|"unboxing"|"fan mail"|"id...   1014651   127794   
2  racist superman|"rudy"|"mancuso"|"king"|"bach"...   3191434   146035   
3  ryan|"higa"|"higatv"|"nigahiga"|"i dare you"|"...   2095828   132239   
4  edsheeran|"ed sheeran"|"acoustic"|"live"|"cove...  33523622  1634130   

   dislikes  comment_count                                  thumbnail_link  \
0     43420         125882  https://i.ytimg.com/vi/n1WpP7iowLc/default.jpg   
1      1688          13030  https://i.ytimg.com/vi/0dBIkQ4Mz1M/default.jpg   
2      5339           8181  https://i.ytimg.com/vi/5qpjK5DgCt4/default.jpg   
3      1989          17518  https://i.ytimg.com/vi/d380meD0W0M/default.jpg   
4     21082          85067  https://i.ytimg.com/vi/2Vv-BfVoq4g/default.jpg   

   comments_disabled  ratings_disabled  video_error_or_removed  \
0              False             False                   False   
1              False             False                   False   
2              False             False                   False   
3              False             False                   False   
4              False             False                   False   

                                         description country  
0  Eminem's new track Walk on Water ft. Beyoncé i...      CA  
1  STill got a lot of packages. Probably will las...      CA  
2  WATCH MY PREVIOUS VIDEO ▶ \n\nSUBSCRIBE ► http...      CA  
3  I know it's been a while since we did this sho...      CA  
4  🎧: https://ad.gt/yt-perfect\n💰: https://atlant...      CA

In [5]:
df_all.describe()

category_id         views         likes      dislikes  comment_count
count  375942.000000  3.759420e+05  3.759420e+05  3.759420e+05   3.759420e+05
mean       20.232302  1.326568e+06  3.788431e+04  2.126107e+03   4.253775e+03
std         7.132413  7.098568e+06  1.654131e+05  2.248437e+04   2.545876e+04
min         1.000000  1.170000e+02  0.000000e+00  0.000000e+00   0.000000e+00
25%        17.000000  4.697800e+04  6.690000e+02  4.100000e+01   1.090000e+02
50%        23.000000  1.773705e+05  3.446000e+03  1.790000e+02   5.110000e+02
75%        24.000000  6.476792e+05  1.747650e+04  7.490000e+02   2.011000e+03
max        44.000000  4.245389e+08  5.613827e+06  1.944971e+06   1.626501e+06

In [6]:
df_all_d = df_all.drop_duplicates() # dropping all duplicates
print(f"""Initial dataset dimensions: {df_all.shape}\n 
Dataset with no duplicates dimensions: {df_all_d.shape}\n
Number of duplicate rows: {df_all.shape[0]-df_all_d.shape[0]}""") #12570 rows duplicate

Initial dataset dimensions: (375942, 17)
 
Dataset with no duplicates dimensions: (363372, 17)

Number of duplicate rows: 12570


In [7]:
df_all_d = df_all_d.copy()

In [8]:
df_all_d.isnull().sum() #18681 nulls in description

video_id                      0
trending_date                 0
title                         0
channel_title                 0
category_id                   0
publish_time                  0
tags                          0
views                         0
likes                         0
dislikes                      0
comment_count                 0
thumbnail_link                0
comments_disabled             0
ratings_disabled              0
video_error_or_removed        0
description               18681
country                       0
dtype: int64

----------------------------------------------

#### 2. Extract all videos that have no tag.

In [9]:
conteggio = df_all_d["tags"].value_counts()
print(conteggio[conteggio > 1])

tags
[none]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           36221
шарий|"шарий блог"|"новое видео шарий"|"шарий новое"|"анатолий шарий"|"шарий анатолий"|"ifhbq"|"видео шария"|"толя шарий"                                                                                                                                                                                                                                                                                                                                                                                   

In [10]:
df_notag = df_all_d[df_all_d['tags'] == '[none]']
print("The rows with the tag '[none]' are:", len(df_notag)) #36221 no tag rows

The rows with the tag '[none]' are: 36221


--------------------------

#### 3. For each channel, determine the total number of views

In [11]:
df_all_d.groupby(df_all_d["channel_title"])[["views"]].sum().reset_index().sort_values("views", ascending=[False])[:10] #group by channel title (as a column not an index), sum the views and order

channel_title        views
4564    ChildishGambinoVEVO  10900185104
15536  Marvel Entertainment  10120133557
17726            NickyJamTV   9479859505
18466                 Ozuna   8623329509
28412               ibighit   7644304297
6689              DrakeVEVO   7637228580
2788              Bad Bunny   7124207494
2101       ArianaGrandeVevo   6202230488
28621      jypentertainment   5790924646
7047             Ed Sheeran   5720351497

-----------------------

#### 4. Save all rows with disabled comments and disabled ratings, or that have video_error_or_removed in a new dataframe called excluded, and remove those rows from the original dataframe.

In [12]:
condizione = ((df_all_d["comments_disabled"] == True) & (df_all_d["ratings_disabled"] == True)) | (df_all_d["video_error_or_removed"] == True) #True è ridondante
excluded = df_all_d[condizione]
df_all_e= df_all_d[~condizione]

In [13]:
df_all_e = df_all_e.copy()

In [14]:
print(f"""Initial dataset dimensions: {df_all_d.shape}\n
Dataset with no excluded dimensions: {df_all_e.shape}\n
Excluded dataset dimensions: {excluded.shape}""")

Initial dataset dimensions: (363372, 17)

Dataset with no excluded dimensions: (360934, 17)

Excluded dataset dimensions: (2438, 17)


-----------------------

#### 5. Add a like_ratio column storing the ratio between the number of likes and of dislikes

In [15]:
df_all_e[["likes", "dislikes"]].eq(0).sum() # we have many 0 in like and dislike that will cause problems

likes       4643
dislikes    6974
dtype: int64

In [16]:
df_all_e["like_ratio"] = df_all_e["likes"] / df_all_e["dislikes"] 
pd.Series({
    "inf": np.isinf(df_all_e["like_ratio"]).sum(), # inf happens when dislike = 0 and like is not 0
    "nan": df_all_e["like_ratio"].isna().sum() # nan happens when like and dislike ara 0
})

inf    2355
nan    4619
dtype: int64

In [17]:
df_all_norm = df_all_e.copy()
df_all_norm["like_ratio"] = df_all_norm["likes"] / (df_all_norm["likes"] + df_all_norm["dislikes"]) # to have no inf and nan we normalize the ratio
df_all_norm["like_ratio"].value_counts()

like_ratio
1.000000    2355
0.800000     283
0.875000     279
0.857143     256
0.900000     248
            ... 
0.943302       1
0.925598       1
0.962228       1
0.964374       1
0.626394       1
Name: count, Length: 238073, dtype: int64

--------------------

#### 6. Cluster the publish time into 10-minute intervals (e.g. from 02:20 to 02:30)

In [18]:
df_all_norm["publish_time"] = pd.to_datetime(df_all_norm["publish_time"]) # changing the type to data

In [19]:
df_all_norm["time_interval_start"] = df_all_norm["publish_time"].dt.floor("10min").dt.strftime("%H:%M") # strftime("%H:%M") extract only the time and not the date
                                                                                                  # floor("10min") returns the closest time interval (17:19:30 -> 17:10)

In [20]:
df_all_norm["time_interval_start"].unique() # checking the time intervals

array(['17:00', '19:00', '18:00', '11:00', '07:30', '23:50', '17:10',
       '20:10', '14:10', '15:00', '16:00', '16:40', '22:00', '01:30',
       '14:00', '12:20', '21:50', '02:00', '00:20', '20:00', '18:20',
       '23:00', '15:50', '04:20', '00:40', '14:50', '19:30', '03:00',
       '22:50', '08:50', '05:50', '01:10', '15:30', '17:50', '05:30',
       '21:10', '06:10', '15:40', '18:30', '12:40', '16:20', '04:40',
       '11:20', '02:10', '16:50', '19:50', '04:30', '22:10', '13:30',
       '05:40', '07:10', '04:50', '13:40', '05:10', '11:30', '03:10',
       '21:40', '17:40', '21:20', '02:30', '23:40', '05:20', '06:40',
       '17:30', '07:00', '20:30', '23:20', '08:30', '00:00', '13:50',
       '13:20', '22:20', '05:00', '02:40', '16:10', '08:00', '17:20',
       '09:20', '11:50', '23:30', '11:10', '16:30', '01:40', '12:50',
       '14:30', '19:20', '21:00', '22:30', '13:00', '19:10', '18:10',
       '06:20', '14:40', '12:00', '20:40', '15:10', '21:30', '22:40',
       '15:20', '01:

In [21]:
df_all_norm["time_interval_end"] = (
    pd.to_datetime(df_all_norm["time_interval_start"], format="%H:%M")
    + pd.Timedelta(minutes=10)
).dt.strftime("%H:%M")

In [22]:
df_all_norm.head()

video_id trending_date  \
0  n1WpP7iowLc      17.14.11   
1  0dBIkQ4Mz1M      17.14.11   
2  5qpjK5DgCt4      17.14.11   
3  d380meD0W0M      17.14.11   
4  2Vv-BfVoq4g      17.14.11   

                                               title channel_title  \
0         Eminem - Walk On Water (Audio) ft. Beyoncé    EminemVEVO   
1                      PLUSH - Bad Unboxing Fan Mail     iDubbbzTV   
2  Racist Superman | Rudy Mancuso, King Bach & Le...  Rudy Mancuso   
3                           I Dare You: GOING BALD!?      nigahiga   
4        Ed Sheeran - Perfect (Official Music Video)    Ed Sheeran   

   category_id              publish_time  \
0           10 2017-11-10 17:00:03+00:00   
1           23 2017-11-13 17:00:00+00:00   
2           23 2017-11-12 19:05:24+00:00   
3           24 2017-11-12 18:01:41+00:00   
4           10 2017-11-09 11:04:14+00:00   

                                                tags     views    likes  \
0  Eminem|"Walk"|"On"|"Water"|"Aftermath/Shady/In...  17158579   787425   
1  plush|"bad unboxing"|"unboxing"|"fan mail"|"id...   1014651   127794   
2  racist superman|"rudy"|"mancuso"|"king"|"bach"...   3191434   146035   
3  ryan|"higa"|"higatv"|"nigahiga"|"i dare you"|"...   2095828   132239   
4  edsheeran|"ed sheeran"|"acoustic"|"live"|"cove...  33523622  1634130   

   dislikes  comment_count                                  thumbnail_link  \
0     43420         125882  https://i.ytimg.com/vi/n1WpP7iowLc/default.jpg   
1      1688          13030  https://i.ytimg.com/vi/0dBIkQ4Mz1M/default.jpg   
2      5339           8181  https://i.ytimg.com/vi/5qpjK5DgCt4/default.jpg   
3      1989          17518  https://i.ytimg.com/vi/d380meD0W0M/default.jpg   
4     21082          85067  https://i.ytimg.com/vi/2Vv-BfVoq4g/default.jpg   

   comments_disabled  ratings_disabled  video_error_or_removed  \
0              False             False                   False   
1              False             False                   False   
2              False             False                   False   
3              False             False                   False   
4              False             False                   False   

                                         description country  like_ratio  \
0  Eminem's new track Walk on Water ft. Beyoncé i...      CA    0.947740   
1  STill got a lot of packages. Probably will las...      CA    0.986963   
2  WATCH MY PREVIOUS VIDEO ▶ \n\nSUBSCRIBE ► http...      CA    0.964730   
3  I know it's been a while since we did this sho...      CA    0.985182   
4  🎧: https://ad.gt/yt-perfect\n💰: https://atlant...      CA    0.987263   

  time_interval_start time_interval_end  
0               17:00             17:10  
1               17:00             17:10  
2               19:00             19:10  
3               18:00             18:10  
4               11:00             11:10

--------------------------

#### 7. For each interval, determine the number of videos, average number of likes and of dislikes.

In [24]:
df_all_norm.groupby("time_interval_start").agg(tot_video=("video_id", "count"), mean_likes=("likes", "mean"), mean_dislikes=("dislikes", "mean")).reset_index()[:10]


time_interval_start  tot_video     mean_likes  mean_dislikes
0               00:00       2815   62978.858615    3910.595027
1               00:10       1468   23091.891689    1482.864441
2               00:20       1216   21799.714638    1092.886513
3               00:30       1552   38174.730026     977.468428
4               00:40       1227   43609.806031    1971.171149
5               00:50       1284   38478.238318    1623.239875
6               01:00       2647   16426.080091     755.600680
7               01:10       1349  106895.709414    5938.600445
8               01:20       1442   25672.203883    1728.871012
9               01:30       1584   13922.699495    6047.232323

---------------------

#### 8. For each tag, determine the number of videos
Notice that tags contains a string with several tags.

In [ ]:
diz_tag = {}

for riga in df_all_norm["tags"]:
    tags = riga.split("|") # split the string in the different tags
    
    for tag in tags:
        nome_tag = tag.strip().strip('"')
        
        if nome_tag == "" or nome_tag == "[none]": # we skipp [none] as it is not a real tag
            continue
        
        if nome_tag in diz_tag:
            diz_tag[nome_tag] += 1
        else:
            diz_tag[nome_tag] = 1

diz_tag

{'Eminem': 461,
 'Walk': 139,
 'On': 96,
 'Water': 108,
 'Aftermath/Shady/Interscope': 127,
 'Rap': 2293,
 'plush': 19,
 'bad unboxing': 20,
 'unboxing': 1259,
 'fan mail': 39,
 'idubbbztv': 38,
 'idubbbztv2': 33,
 'things': 105,
 'best': 1958,
 'packages': 13,
 'plushies': 22,
 'chontent chop': 9,
 'racist superman': 119,
 'rudy': 259,
 'mancuso': 210,
 'king': 188,
 'bach': 94,
 'racist': 120,
 'superman': 263,
 'love': 2141,
 'rudy mancuso poo bear black white official music video': 31,
 'iphone x by pineapple': 13,
 'lelepons': 674,
 'hannahstocking': 656,
 'rudymancuso': 656,
 'inanna': 588,
 'anwar': 646,
 'sarkis': 582,
 'shots': 686,
 'shotsstudios': 637,
 'alesso': 686,
 'anitta': 702,
 'brazil': 631,
 "Getting My Driver's License": 16,
 'Lele Pons': 176,
 'ryan': 510,
 'higa': 193,
 'higatv': 209,
 'nigahiga': 251,
 'i dare you': 13,
 'idy': 13,
 'rhpc': 17,
 'dares': 21,
 'no truth': 13,
 'comments': 104,
 'comedy': 11968,
 'funny': 14573,
 'stupid': 121,
 'fail': 1517,
 'ed

------------------------------

#### 9. Find the tags with the largest number of videos

In [ ]:
top_10 = sorted(diz_tag.items(), key=lambda x: x[1], reverse=True)[:10]
top_10

[('funny', 14573),
 ('comedy', 11968),
 ('2018', 11224),
 ('news', 6214),
 ('music', 5845),
 ('2017', 5652),
 ('video', 5536),
 ('humor', 5026),
 ('review', 4021),
 ('television', 4011)]

-------------------------------

#### 10. For each (tag, country) pair, compute average ratio likes/dislikes

In [ ]:
df_all_tag = df_all_norm.copy()
df_all_tag = df_all_tag[["tags", "country", "like_ratio"]]

In [ ]:
df_all_tag["tags_list"] = df_all_tag["tags"].str.split("|")
df_tags = df_all_tag.explode("tags_list") # creates a row for each tag after the split
df_tags = df_tags.replace(to_replace=["","[none]"], value=np.nan)

In [ ]:
df_result = df_tags.groupby(["tags_list", "country"])[["like_ratio"]].mean().reset_index()
df_result = df_result.rename(columns = {"like_ratio" : "mean_like_ratio"})
df_result.head()

tags_list country  mean_like_ratio
0      #Freeticket"      IN         0.727197
1             #GST       IN         0.707337
2        #Jaisimha       IN         0.727197
3   #JanaSenaParty       IN         0.946263
4       #MahaaNews"      IN         0.894829

----------------------------

#### 11. For each (trending_date, country) pair, the video with the largest number of views

In [ ]:
df_all_norm["trending_date"] = pd.to_datetime(df_all_norm["trending_date"], format="%y.%d.%m")

In [ ]:
id_max = df_all_norm.groupby(["trending_date", "country"])["views"].idxmax()
df_all_norm.loc[id_max, ["trending_date", "country", "video_id", "title", "views"]][:10]

trending_date country     video_id  \
4         2017-11-14      CA  2Vv-BfVoq4g   
40912     2017-11-14      DE  2Vv-BfVoq4g   
81895     2017-11-14      FR  2Vv-BfVoq4g   
122451    2017-11-14      GB  2Vv-BfVoq4g   
161378    2017-11-14      IN  ePO5M5DE01I   
219394    2017-11-14      KR  n1WpP7iowLc   
253980    2017-11-14      MX  JzCsM1vtn78   
294429    2017-11-14      RU  U_4BLnJuR3g   
335063    2017-11-14      US  2Vv-BfVoq4g   
282       2017-11-15      CA  2Vv-BfVoq4g   

                                                    title     views  
4             Ed Sheeran - Perfect (Official Music Video)  33523622  
40912         Ed Sheeran - Perfect (Official Music Video)  33523622  
81895         Ed Sheeran - Perfect (Official Music Video)  33523622  
122451        Ed Sheeran - Perfect (Official Music Video)  33523622  
161378  Tiger Zinda Hai | Official Trailer | Salman Kh...  35885754  
219394         Eminem - Walk On Water (Audio) ft. Beyoncé  17158579  
253980               THE LOGANG MADE HISTORY. LOL. AGAIN.   4477587  
294429            12 лайфхаков с термоклеем для хендмейда   2109573  
335063        Ed Sheeran - Perfect (Official Music Video)  33523622  
282           Ed Sheeran - Perfect (Official Music Video)  39082222

----------------------------------

#### 12. Divide trending_date into three columns: year, month, day

In [ ]:
df_all_norm["trending_year"] = df_all_norm["trending_date"].dt.year
df_all_norm["trending_month"] = df_all_norm["trending_date"].dt.month
df_all_norm["trending_day"] = df_all_norm["trending_date"].dt.day

-------------------------

#### 13. For each (month, country) pair, the video with the largest number of views

In [ ]:
id_max_mc = df_all_norm.groupby(["trending_month", "country"])["views"].idxmax()
df_all_norm.loc[id_max_mc, ["trending_month", "country", "video_id", "title", "views"]][:10]

trending_month country     video_id  \
11214                1      CA  LsoLEjrDogU   
51930                1      DE  LsoLEjrDogU   
92853                1      FR  LsoLEjrDogU   
135208               1      GB  LsoLEjrDogU   
173405               1      IN  dfnCAmr569k   
228682               1      KR  LsoLEjrDogU   
264769               1      MX  LsoLEjrDogU   
306098               1      RU  dfnCAmr569k   
346578               1      US  LsoLEjrDogU   
20055                2      CA  xpVfcZ0ZcFM   

                                                    title     views  
11214   Bruno Mars - Finesse (Remix) [Feat. Cardi B] [...  43067983  
51930   Bruno Mars - Finesse (Remix) [Feat. Cardi B] [...  37728802  
92853   Bruno Mars - Finesse (Remix) [Feat. Cardi B] [...  37728802  
135208  Bruno Mars - Finesse (Remix) [Feat. Cardi B] [...  90598955  
173405     Taylor Swift - End Game ft. Ed Sheeran, Future  42019590  
228682  Bruno Mars - Finesse (Remix) [Feat. Cardi B] [...  37728802  
264769  Bruno Mars - Finesse (Remix) [Feat. Cardi B] [...  31680160  
306098     Taylor Swift - End Game ft. Ed Sheeran, Future  23198594  
346578  Bruno Mars - Finesse (Remix) [Feat. Cardi B] [...  57951412  
20055                                  Drake - God’s Plan  47362934

------------------------------

#### 14. Read all json files with the video categories

In [ ]:
countries = ["CA", "DE", "FR", "GB", "IN", "JP", "KR", "MX", "RU", "US"]
dfs2 = {}

for element in countries:
    path = fr"C:\Users\chiar\OneDrive\Desktop\foundation of computer science\trendingYT\{element}_category_id.json"
    print("Reading:", path)

    with open(path, "r", encoding="utf-8") as f:
        data = json.load(f)
        df_json = pd.json_normalize(data["items"])
        
        df_json["country"] = element
        
        # Salva il DataFrame nel dizionario
        dfs2[element] = df_json

Reading: C:\Users\chiar\OneDrive\Desktop\foundation of computer science\trendingYT\CA_category_id.json
Reading: C:\Users\chiar\OneDrive\Desktop\foundation of computer science\trendingYT\DE_category_id.json
Reading: C:\Users\chiar\OneDrive\Desktop\foundation of computer science\trendingYT\FR_category_id.json
Reading: C:\Users\chiar\OneDrive\Desktop\foundation of computer science\trendingYT\GB_category_id.json
Reading: C:\Users\chiar\OneDrive\Desktop\foundation of computer science\trendingYT\IN_category_id.json
Reading: C:\Users\chiar\OneDrive\Desktop\foundation of computer science\trendingYT\JP_category_id.json
Reading: C:\Users\chiar\OneDrive\Desktop\foundation of computer science\trendingYT\KR_category_id.json
Reading: C:\Users\chiar\OneDrive\Desktop\foundation of computer science\trendingYT\MX_category_id.json
Reading: C:\Users\chiar\OneDrive\Desktop\foundation of computer science\trendingYT\RU_category_id.json
Reading: C:\Users\chiar\OneDrive\Desktop\foundation of computer science\t

------------------

#### 15: For each country, determine how many videos have a category that is not assignable.

In [ ]:
df_all_json = pd.concat(dfs2.values(), ignore_index=True) # concatenation of all datasets from the dictionary
df_all_json.head()

kind                                               etag  \
0  youtube#videoCategory  "ld9biNPKjAjgjV7EZ4EKeEGrhao/Xy1mB4_yLrHy_BmKm...   
1  youtube#videoCategory  "ld9biNPKjAjgjV7EZ4EKeEGrhao/UZ1oLIIz2dxIhO45Z...   
2  youtube#videoCategory  "ld9biNPKjAjgjV7EZ4EKeEGrhao/nqRIq97-xe5XRZTxb...   
3  youtube#videoCategory  "ld9biNPKjAjgjV7EZ4EKeEGrhao/HwXKamM1Q20q9BN-o...   
4  youtube#videoCategory  "ld9biNPKjAjgjV7EZ4EKeEGrhao/9GQMSRjrZdHeb1OEM...   

   id         snippet.channelId     snippet.title  snippet.assignable country  
0   1  UCBR8-60-B28hp2BmDPdntcQ  Film & Animation                True      CA  
1   2  UCBR8-60-B28hp2BmDPdntcQ  Autos & Vehicles                True      CA  
2  10  UCBR8-60-B28hp2BmDPdntcQ             Music                True      CA  
3  15  UCBR8-60-B28hp2BmDPdntcQ    Pets & Animals                True      CA  
4  17  UCBR8-60-B28hp2BmDPdntcQ            Sports                True      CA

In [ ]:
df_all_json["snippet.assignable"] = df_all_json["snippet.assignable"].astype(bool)
print(df_all_json["snippet.assignable"].value_counts())

snippet.assignable
False    170
True     141
Name: count, dtype: int64


In [ ]:
print(df_all_json["id"].dtype)
df_all_json["id"] = df_all_json["id"].astype(int)
print(df_all_json["id"].dtype)

object
int32


In [ ]:
df_all_json = df_all_json.rename(columns={"id": "category_id"}) # rename to prepare for a merge
merged = pd.merge(df_all_norm, df_all_json, on = ["category_id", "country"] , how="left" )

In [ ]:
merged[merged["snippet.assignable"] == False].groupby(["country"])[["video_id"]].count().reset_index()

country  video_id
0      CA       130
1      DE       110
2      FR       112
3      GB        20
4      IN       159
5      KR       159
6      MX         3
7      RU       194
8      US        57

In [ ]:
merged["snippet.assignable"].value_counts(dropna=False)

snippet.assignable
True     357333
NaN        2657
False       944
Name: count, dtype: int64

da ricontrollare perchè ci sono 2657 righe in cui non è stato assegnato un valore per snippet.assignable durante il merge